### PR TITLE
Feat: Simple Code Motion

### PR DESCRIPTION
**Pull request by mahaloz**
## Desc
Given two blocks with similar statements that overlap, try to move those statements out of the block and condense them. This is very similar to normal compiler code motion optimization. 

## Summary of Changes
- added 2 new files for similarity and io detection AIL Blocks 
- added optimization pass to move similar code out of blocks
- 1 new test case

### Total Test Time 
- Before Changes: `real    56m51.998s`
- After Changes: `real    54m0.059s`

No I don't understand why it's faster now... 

## Example
Turns:
```c
    if (x) {
        b = 2;
        a = 1;
        c = 3;
    } else {
        b = 2;
        c = 3;
    }
```
Into:
```c
    if (x) {
        a = 1;
    }
    b = 2;
    c = 3;
```

## TODO
- [x] Finish the ends-movement based on IOFinder
- [x] Test on all slowtests 
- [x] Document code more
- [x] Make issues for TODOs

### CODE REVIEW

**Comment by mahaloz** [comment]
## Demo

https://github.com/angr/angr/assets/21327264/eaff0843-bca1-479d-beca-871ab1f4ef16

**Comment by mahaloz** [comment]
Waiting on confirmation from @AOS002 before merging this (after tests pass).

File: angr/analyses/decompiler/block_io_finder.py
Comment (ltfish) [inline angr/analyses/decompiler/block_io_finder.py:92]: `NotImplementedError` means you'll implement it in the future. I suppose that's not what you mean here?

File: angr/analyses/decompiler/block_io_finder.py
Comment (ltfish) [inline angr/analyses/decompiler/block_io_finder.py:117]: Same here. I think you almost want to raise a `RuntimeError` or an `AngrError` here. We generally prefer the latter so that it's catchable.

File: angr/analyses/decompiler/optimization_passes/code_motion.py
Comment (ltfish) [inline angr/analyses/decompiler/optimization_passes/code_motion.py:48]: How about "Merge common statements in sub scopes"?

File: angr/analyses/decompiler/utils.py
Comment (ltfish) [inline angr/analyses/decompiler/utils.py:738]: Defining functions differently in different branches is a bad programming practice. Please refactor it.

File: tests/analyses/decompiler/test_decompiler.py
Comment (ltfish) [inline tests/analyses/decompiler/test_decompiler.py:532]: I believe doc strings in test cases will be used in the CI output instead of the function name. That's why we have been avoiding having doc strings for test cases. Can you check if it's still the case with CI?

File: tests/analyses/decompiler/test_decompiler.py
Comment (ltfish) [inline tests/analyses/decompiler/test_decompiler.py:3366]: I guess you forget to specify a structuring algorithm for this test case.

**Review by ltfish — CHANGES_REQUESTED** [review]

File: tests/analyses/decompiler/test_decompiler.py
Comment (mahaloz) [reply tests/analyses/decompiler/test_decompiler.py:532]: @ltfish how can I check this? I've been adding docstrings into the tests for a while now... and I don't see them in the output at least on GitHub.

Comment (mahaloz) [reply tests/analyses/decompiler/test_decompiler.py:532]: Yeah I dont see it used in the CI

**Review by ltfish — APPROVED** [review]

**Pull request merged**
